### DISCUSSION TITLE
❗🛠 Testers, documentors needed

### DISCUSSION THREAD

**Discussion by demberto** [demberto/PyFLP#19] [Announcements]
✨An upcoming version of PyFLP is coming soon #15.

🛠 I need testers to test their own FLPs against the latest version, which can be installed via cloning the repo and performing an editable install.

📘 I have uploaded the latest docs to <https://pyflp.rtfd.io>.

**Comment by ttaschke** [comment]
Hi, I would like to help with this. 
Please let me know, what kind of tests / feedback you are looking for.

For now I just tried to import a random FLP, but got the following parse error:
```
Python 3.10.7 (main, Sep  7 2022, 22:40:42) [GCC 11.2.1 20220219] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyflp
>>> project = pyflp.parse("../test_data/melody.flp")
/python_project/pyflp/pyflp/__init__.py:151: RuntimeWarning: Event 212 not parsed entirely; parsed 18, found 52 bytes
  events.append(event_type(id, value))
/python_project/pyflp/pyflp/__init__.py:151: RuntimeWarning: Event 215 not parsed entirely; parsed 158, found 168 bytes
  events.append(event_type(id, value))
```
(Note: Above was executed after applying the fix for Python 3.10 from https://github.com/demberto/PyFLP/pull/21)

I also tried an older Python version, but got an error on importing pyflp:
```
Python 3.7.14 (default, Sep  8 2022, 00:44:25)
[GCC 11.2.1 20220219] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyflp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/python_project/pyflp/pyflp/__init__.py", line 37, in <module>
    from ._base import (
  File "/python_project/pyflp/pyflp/_base.py", line 93, in <module>
    class EventBase(abc.ABC, Generic[T], Hashable, Sized, SupportsBytes):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

-----

Btw, I would also be happy to help with development tasks, if you could suggest some appropriate development tasks for starting to contribute to this project in general.

**Reply by demberto** [reply]
> For now I just tried to import a random FLP, but got the following parse error

Its a warning, which just tells that an event size has increased compared to what I parse into `pyflp._base.StructBase._props`.

> Hi, I would like to help with this.
Please let me know, what kind of tests / feedback you are looking for.

I need to first get unittests running for all the properties. Check the unittests branch, I have got a few running.

All the properties are parsed lazily, and this is a double edged sword. For one, it doesn't give up because some new version of FLP breaks the existing parsing logic itself. I get rid of private variables which were a massive PITA and were a reason for all the spaghetti code in 1.x.

So unless, each one of them is tested, we have no way to find if its logic is broken. Now this won't happen for simple events, it will be mostly a thing when the  gathers events required for a model.

Super old versions of FLP are way more messed up in terms of ordering of events. Its literally as if the history of changes were converted to events and dumped. Most of these IDs are deprecated and some are even probably repurposed in newer versions.

I'd begin with opening the FLP in FL Studio and compare values from the interface to those in the parsed `Project`. I have a useful utility for looking at events in an FLP, its what FLPInspect aimed to replace, its Windows only though.

> Btw, I would also be happy to help with development tasks, if you could suggest some appropriate development tasks.

If you are familiar with Sphinx, you can take a look at the docs, there is a branch for it as well. Unlike other libraries, PyFLP's docs are purely based in the API reference. I want to make them look as intuitive as they can, by using tabs, grids, cards, admonitions etc. I plan to document them in a way familiar to FL Studio's manual. That's why I have used images. Docstrings are also a problem, see #17.

If you like reverse engineering stuff, there are tons of undiscovered fields in variable events. I have a tool called FLPEdit, which makes this way easier. `ChannelID.Parameters` is one such event (its 168 bytes!). I have prefixed undiscovered fields with `_u`. Support for automation and controllers is non existent, I am not even sure if the current remote controller code is correct. You can also implement parsers for native plugin data, check `pyflp.plugin`.

If you'd like to improve the representation of existing properties, take a look at all "length" and "position" properties. These all are PPQ dependant. They can be changed to return custom "Length" and "Position" objects, which convert integers into beats and bars. If the PPQ currently is 96, and the length of something is 96 it means that thing is one beat / 1/4th bar. However, the real challenge is with non 4/4 time signatures. I have no idea how musical times work, so I have nothing further to say.

There are TODO comments at a lot of places in the code, so you could just browse the code and find them

**Reply by ttaschke** [reply]
Thanks for the explanation. 
I will take a look at the things you mentioned and the TODOs in code and see what makes sense for first steps.

> Its a warning, which just tells that an event size has increased compared to what I parse into pyflp._base.StructBase._props.

Oh, right.. Totally overlooked that it is a warning and not an error.



Also, if you do not have this on your radar yet, I will create an issue for the error with Python 3.7 and look into it.
`TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`

**Reply by demberto** [reply]
> Also, if you do not have this on your radar yet, I will create an issue for the error with Python 3.7 and look into it.
> `TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`

I found out that inheriting `typing.SupportsBytes` for `EventBase` did that, fixed it.

**Reply by demberto** [reply]
@ttaschke Hey there, I have opened some issues. Please check them out

**Reply by ttaschke** [reply]
Hi @demberto,
While looking through the issues, I got interested in seeing how inspecting FLPs with your GUI application works.
I could not get it to run immediately, because it depends on an old version of the PyFLP API, but I got it working with the newest PyFLP version in the end.
I also created a PR for this:
https://github.com/demberto/FLPInspect/pull/1
(In case it is too early to sync FLPInspect to PyFLP, this can maybe just be the base for an upgrade after the full 2.x release.)

Btw, originally I was looking for how to do inspecting and finding events. Could you please explain what your workflow for this is?
For example, when trying to map a plugin's params to events (https://github.com/demberto/PyFLP/issues/30).

**Reply by demberto** [reply]
Nah FLPInspect was just a side hustle, I will be working on a rewrite of it after I finish with PyFLP and the ZIP export utility. There was a utility called FLPEdit by @roadcrewworker. I dunno why it got taken down. It's a nice utlity to inspect event data. If you are on WIndows you are in luck, I have attached it.

Go to `FLPEdit/bin/release`, and check the **DBG: Allow unknown IDs**. A good amount of it's code was used for PyFLP's event parsing logic. It might run just fine if you are on Mac or Linux via Wine.

![image](https://user-images.githubusercontent.com/65948403/190705441-574d76a6-eaff-4974-bd60-6c095bddff64.png)

[FLPEdit.zip](https://github.com/demberto/PyFLP/files/9586342/FLPEdit.zip)

**Comment by demberto** [comment]
@ttaschke I have added new docs for users and contributors, can you check them out please? 👀

**Reply by ttaschke** [reply]
Sorry, only noticed this now.
Just had a quick look at the contributor guide and this seems really comprehensive. 👍 🙏 
A lot of information there so it will take a bit time to work through it all.

**Comment by demberto** [comment]
It just occured to me that instead of using the same FLP for all the tests, separate presets (saved via FL Studio) can be used for separate tests. Modifying them is as easy as loading the preset in FL and saving it back.

**Presets saved in one of FL Studio's formats use the same layout as a full FLP, but only the required events**.

This will also improve test isolation.
